staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -596,27 +618,27 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, pred stor
 			growSlice(v, 2048, len(getResp.Kvs))
 		}
 
-		// take items from the response until the bucket is full, filtering as we go
-		for _, kv := range getResp.Kvs {
-			if paging && int64(v.Len()) >= pred.Limit {
-				hasMore = true
-				break
-			}
-			lastKey = kv.Key
-
-			data, _, err := s.transformer.TransformFromStorage(kv.Value, authenticatedDataString(kv.Key))
-			if err != nil {
-				return storage.NewInternalErrorf("unable to transform key %q: %v", kv.Key, err)
-			}
+		if paging && pred.Limit < int64(len(getResp.Kvs)) {
+			getResp.Kvs = getResp.Kvs[0:pred.Limit]
+			hasMore = true
+		}
```

Comment:
I do not think this block is ever taken.  etcd should never return more data than what you asked for via the limit.

Reply:
Removed.